### PR TITLE
Annotated the text package

### PR DESCRIPTION
**Pull request by Kakarot-SSJ4**

### CODE REVIEW

File: Stubs/CharBuffer.astub [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import org.checkerframework.checker.index.qual.NonNegative;
+
+package java.nio;
+
+class CharBuffer extends Buffer implements Comparable<CharBuffer>, Appendable, CharSequence, Readable
```
Comment (kelloggm) [inline Stubs/CharBuffer.astub:5]: Why not add this class to the Index annotated JDK? `java.nio.Buffer` and `java.nio.ByteBuffer` are already there, so it should be pretty straightforward to check that the annotations are consistent.

File: Stubs/ParsePosition.astub [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import org.checkerframework.checker.index.qual.NonNegative;
+package java.text;
+
+class ParsePosition {
```
Comment (kelloggm) [inline Stubs/ParsePosition.astub:4]: I would also add this class to the annotated JDK.

Comment (kelloggm) [reply Stubs/ParsePosition.astub:4]: Stubs are good for early development work, but when you want a proof like this to rely on them it's better to share that work with others.

File: pom.xml [hunk 3]
```diff
@@ -550,6 +550,17 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.checkerframework</groupId>
+      <artifactId>checker-qual</artifactId>
+      <version>2.8.1</version>
```
Comment (kelloggm) [inline pom.xml]: Could change this to 2.8.2, since that's now the latest version.

File: pom.xml [hunk 4]
```diff
@@ -888,6 +906,50 @@
...
+                </path>
+              </annotationProcessorPaths>
+              <annotationProcessors>
+                <!-- Add all the checkers you want to enable here -->
+                <annotationProcessor>org.checkerframework.checker.index.IndexChecker</annotationProcessor>
+              </annotationProcessors>
+              <compilerArgs>
+                <!-- location of the annotated JDK, which comes from a Maven dependency -->
+                <arg>-Xbootclasspath/p:${annotatedJdk}</arg>
+                <!-- -Awarns turns type-checking warnings into errors. -->
```
Comment (kelloggm) [inline pom.xml]: If you don't intend to include this option, why comment it out instead of removing it?

File: pom.xml [hunk 5]
```diff
@@ -844,6 +857,11 @@
           </tagListOptions>
         </configuration>
       </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
```
Comment (kelloggm) [inline pom.xml]: That is the purpose of adding this plugin? The Checker Framework's Maven instructions (https://checkerframework.org/manual/#maven) don't mention it.

File: src/main/java/org/apache/commons/lang3/builder/ToStringStyle.java [hunk 6]
```diff
@@ -436,13 +440,16 @@ public void appendEnd(final StringBuffer buffer, final Object object) {
      * @param buffer  the <code>StringBuffer</code> to populate
      * @since 2.0
      */
+    @SuppressWarnings("index") /* #1: i < sepLen.length() and len >= sepLen => i < len.length()
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/builder/ToStringStyle.java]: `sepLen` is an integer, so `sepLen.length()` doesn't make sense.

Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/builder/ToStringStyle.java]: I think an easier explanation for this would be that `len` is the length of `buffer` and `sepLen` is the length of `fieldSeparator`, and that `len >= sepLen` is checked just above. The reader can figure out `i`'s relationship to `sepLen` very easily by reading the for loop.

File: src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java [hunk 7]
```diff
@@ -253,24 +255,30 @@ public static String getRootCauseMessage(final Throwable th) {
...
-        final Throwable throwables[] = getThrowables(throwable);
+        final Throwable @MinLen(1) [] throwables = getThrowables(throwable); // #1
         final int count = throwables.length;
         final List<String> frames = new ArrayList<>();
         List<String> nextTrace = getStackFrameList(throwables[count - 1]);
         for (int i = count; --i >= 0;) {
             final List<String> trace = nextTrace;
             if (i != 0) {
-                nextTrace = getStackFrameList(throwables[i - 1]);
+                nextTrace = getStackFrameList(throwables[i - 1]); // #2
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java]: Shouldn't `i` be positive here? I would expect that everywhere inside the for loop `i`'s type is `@NonNegative`, since that's the condition that's checked. If a non-negative int is not zero, it should be positive.

Can you write a minimal test case for this that shows the weakness in the checker, and then change the explanation comment above to reference that issue? Thanks.

Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java]: Based on the other warnings below, the test case probably looks something like:

```java
void test(@Positive int count) {
    for (int i = count; --i >= 0;) {
         if (i != 0) {
              @Positive int j = i;
         }
         @NonNegative int k = i;
    }
}
```

File: src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java [hunk 8]
```diff
@@ -936,4 +944,4 @@ public static void removeCommonFrames(final List<String> causeFrames, final List
     public ExceptionUtils() {
         super();
     }
-}
+}
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java]: Please don't make unnecessary whitespace changes.

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java [hunk 9]
```diff
@@ -647,6 +651,9 @@ public static BigDecimal toScaledBigDecimal(final String value, final int scale,
      * @return Number created from the string (or null if the input is null)
      * @throws NumberFormatException if the value cannot be converted
      */
+    @SuppressWarnings("argument.type.incompatible")/*
+    #1, #2, #3: !StringUtils.isBlank(str) => str.length != 0
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: `StringUtils.isBlank()` should be annotated with an `@EnsuresMinLenIf` annotation that says that if it returns false, the string is `@MinLen(1)`

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java [hunk 10]
```diff
@@ -933,7 +940,7 @@ public static Long createLong(final String str) {
         }
         return Long.decode(str);
     }
-
+
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: ?

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java [hunk 11]
```diff
@@ -162,19 +165,23 @@ private static float getTotalTransformationCost(final Class<?>[] srcArgs, final
             // When isVarArgs is true, srcArgs and dstArgs may differ in length.
             // There are two special cases to consider:
             final boolean noVarArgsPassed = srcArgs.length < destArgs.length;
+            @SuppressWarnings("index:array.access.unsafe.low") // isVarArgs => srcArgs.length != 0
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:167]: `isVarArgs` could also be annotated with `@EnsuresMinLenIf`

File: src/main/java/org/apache/commons/lang3/reflect/MethodUtils.java [hunk 12]
```diff
@@ -792,6 +807,7 @@ private static int distance(final Class<?>[] classArray, final Class<?>[] toClas
      * @throws NullPointerException if the specified method is {@code null}
      * @since 3.2
      */
+    @SuppressWarnings({"index:compound.assignment.type.incompatible", "index:array.access.unsafe.high"}) // #4: getGenericParameterTypes() returns an array of Types that represent the formal parameter types of the method object, hence all instances of a class will have the same length of the array returned, hence,  method.getGenericParameterTypes().length =  m.getGenericParameterTypes().length = parameterTypes.length and i < parameterTypes.length inside the loop
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/reflect/MethodUtils.java]: This comment would be much easier to read if it was split over multiple lines. In general, it's good style to avoid lines longer than about 140 characters.

File: src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java [hunk 13]
```diff
@@ -483,15 +514,16 @@ private StringBuilder appendQuotedString(final String pattern, final ParsePositi
...
 
-        final int start = pos.getIndex();
+        final @LTEqLengthOf("pattern") int start = pos.getIndex(); // #1
         final char[] c = pattern.toCharArray();
-        final int lastHold = start;
-        for (int i = pos.getIndex(); i < pattern.length(); i++) {
-            if (c[pos.getIndex()] == QUOTE) {
+        final @LTEqLengthOf("c") int lastHold = start;
+        for (@IndexOrHigh("c") int i = pos.getIndex(); i < pattern.length(); i++) { // #2
+            switch (c[pos.getIndex()]) { // #3
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java]: Why change this `if` to a `switch`? I think they're semantically equivalent, but it clutters the diff.

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 14]
```diff
@@ -189,8 +198,9 @@ public StrBuilder setNullText(String nullText) {
      *
      * @return the length
      */
+    @SuppressWarnings("override.return.invalid") // it says required is @LTEqLengthOf("this"), which does not makes sense, it should be @LTEqLengthOf("this.buffer")
     @Override
-    public int length() {
+    public @NonNegative @LTEqLengthOf("this.buffer") int length() {
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: You're right that a warning needs to be suppressed here, but the annotations are wrong. This class is a `CharSequence` implementation, so a programmer should be able to use it like any other `CharSequence`: that is, they shouldn't be exposed to its implementation details.

The correct signature is `@IndexOrHigh("this")`.

You'll have to suppress a warning to connect the implementation detail `this.buffer` to the `this` object. I suggest doing so on a `SameLen` annotation on `this.buffer`, as described here: https://checkerframework.org/manual/#index-annotating-fixed-size

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 15]
```diff
@@ -2062,7 +2201,7 @@ public StrBuilder replaceAll(final StrMatcher matcher, final String replaceStr)
      * Matchers can be used to perform advanced replace behaviour.
      * For example you could write a matcher to replace
      * where the character 'a' is followed by a number.
-     *
+     *BUILD FAILURE
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: ?

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 16]
```diff
@@ -2408,7 +2563,7 @@ public boolean contains(final StrMatcher matcher) {
      * @param ch  the character to find
      * @return the first index of the character, or -1 if not found
      */
-    public int indexOf(final char ch) {
+    public @IndexOrLow("this.buffer") int indexOf(final char ch) {
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: Should be `IndexOrLow("this")` (here and on other external-facing methods)

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java [hunk 17]
```diff
@@ -657,11 +664,12 @@ public boolean replaceIn(final StringBuffer source, final int offset, final int
      * @return true if altered
      * @since 3.2
      */
+    @SuppressWarnings("index:argument.type.incompatible") // #1: source.length() is as required @LTLengthOf(value="source", offset="-1"), can't find the length function in the stub file
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: I don't understand the problem here

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java [hunk 18]
```diff
@@ -795,23 +813,23 @@ private int substitute(final StrBuilder buf, final int offset, final int length,
                     while (pos < bufEnd) {
                         if (substitutionInVariablesEnabled
                                 && (endMatchLen = pfxMatcher.isMatch(chars,
-                                        pos, offset, bufEnd)) != 0) {
+                                        pos, offset, bufEnd)) != 0) { // #3
                             // found a nested variable start
                             nestedVarCount++;
-                            pos += endMatchLen;
+                            pos += endMatchLen;
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: whitespace

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 19]
```diff
@@ -732,7 +746,8 @@ private int readNextToken(final char[] srcChars, int start, final int len, final
      *  immediately after the delimiter, or if end of string found,
      *  then the length of string
      */
-    private int readWithQuotes(final char[] srcChars, final int start, final int len, final StrBuilder workArea,
+    @SuppressWarnings({"index:argument.type.incompatible","index:array.access.unsafe.high.range","index:array.access.unsafe.low"}) // #1, #2, #3, #4, #5: pos < len is checked as the condition for the while loop, and when all the compound assignments happen (#0.1, #0.2, #0.3, #0.4, #0.5), there is a continue statement, hence the loop starts again and pos < len is checked. Also, pos++ can happen only once, either in #1 or in #5 because both are complementary if-else parts, hence srcChars[pos++] is also validin #1 and #5 as pos < len
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: Make this a multi-line comment

File: src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java [hunk 20]
```diff
@@ -108,9 +111,11 @@ public JavaUnicodeEscaper(final int below, final int above, final boolean betwee
      *            a Unicode code point
      * @return the hex string for the given codepoint
      */
+    @SuppressWarnings("override.param.invalid") // the function it is overriding may take any codepoint value
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java]: That seems like a bug, or at least a violation of the principle of least surprise. If the superclass can take any codepoint, then this one ought to take any codepoint, as well.

File: src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java [hunk 21]
```diff
@@ -296,6 +304,7 @@ private void init() {
                 }
                 break;
             case '\'': // literal text
+                @SuppressWarnings("index:argument.type.incompatible") // token has a minimum length 1
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java]: Say how you know the fact, not just the fact

**Review by kelloggm — CHANGES_REQUESTED** [review]
This is looking good, but does still need some work. See my specific comments.

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java (hunk 17 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: In the stub file of StringBuilder that I attained by the following command:
java -cp $CHECKERFRAMEWORK/checker/dist/checker.jar org.checkerframework.framework.stub.StubGenerator java.lang.StringBuilder > StringBuilder.astub

I can't find the length() function to annotate.
The following is the stub file:
```
package java.lang;

class StringBuilder extends AbstractStringBuilder implements Serializable, CharSequence {
    StringBuilder();
    StringBuilder(int arg0);
    StringBuilder(String arg0);
    StringBuilder(CharSequence arg0);
    StringBuilder append(Object arg0);
    StringBuilder append(String arg0);
    StringBuilder append(StringBuffer arg0);
    StringBuilder append(CharSequence arg0);
    StringBuilder append(CharSequence arg0, int arg1, int arg2);
    StringBuilder append(char[] arg0);
    StringBuilder append(char[] arg0, int arg1, int arg2);
    StringBuilder append(boolean arg0);
    StringBuilder append(char arg0);
    StringBuilder append(int arg0);
    StringBuilder append(long arg0);
    StringBuilder append(float arg0);
    StringBuilder append(double arg0);
    StringBuilder appendCodePoint(int arg0);
    StringBuilder delete(int arg0, int arg1);
    StringBuilder deleteCharAt(int arg0);
    StringBuilder replace(int arg0, int arg1, String arg2);
    StringBuilder insert(int arg0, char[] arg1, int arg2, int arg3);
    StringBuilder insert(int arg0, Object arg1);
    StringBuilder insert(int arg0, String arg1);
    StringBuilder insert(int arg0, char[] arg1);
    StringBuilder insert(int arg0, CharSequence arg1);
    StringBuilder insert(int arg0, CharSequence arg1, int arg2, int arg3);
    StringBuilder insert(int arg0, boolean arg1);
    StringBuilder insert(int arg0, char arg1);
    StringBuilder insert(int arg0, int arg1);
    StringBuilder insert(int arg0, long arg1);
    StringBuilder insert(int arg0, float arg1);
    StringBuilder insert(int arg0, double arg1);
    int indexOf(String arg0);
    int indexOf(String arg0, int arg1);
    int lastIndexOf(String arg0);
    int lastIndexOf(String arg0, int arg1);
    StringBuilder reverse();
    String toString();
}

```

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java (hunk 11 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:167]: But which variable's minimum length would it ensure? I'm unable to figure it out.

File: src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java (hunk 8 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java]: I don't understand this diff, the old and the new versions have the exact same '}' without any whitespace change.

Comment (wmdietl) [reply src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java]: If you hover over the red icon you see `No newline at end of file`. So the original had a newline at the end and it was removed in your change.

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java (hunk 17 above)
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: There are two issues here:
1. `StringBuilder` is a mutable sequence, so upperbound annotations cannot apply to it in general (since its length might change, making the annotations invalid)
2. `StringBuilder` doesn't define its own `length` function; instead, it inherits from `CharSequence`

I would suppress this warning by noting that `StringBuilder` is mutable, so you can't be sure of this in general.

**Comment by kelloggm** [comment]
@Kakarot-SSJ4  Why does the build for this fail?

**Comment by Kakarot-SSJ4** [comment]
When I compile it locally, it shows `BUILD SUCCESS`

**Comment by Kakarot-SSJ4** [comment]
@kelloggm @panacekcz

File: pom.xml [hunk 22]
```diff
@@ -550,6 +550,17 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.checkerframework</groupId>
+      <artifactId>checker-qual</artifactId>
+      <version>2.8.2</version>
```
Comment (kelloggm) [inline pom.xml]: Should be updated to 2.9.0

File: src/main/java/org/apache/commons/lang3/math/IEEE754rUtils.java [hunk 23]
```diff
@@ -41,6 +41,7 @@ public static double min(final double... array) {
         Validate.isTrue(array.length != 0, "Array cannot be empty.");
 
         // Finds and returns min
+        @SuppressWarnings("index:array.access.unsafe.high.constant") // Validate.isTrue(array.length != 0, "Array cannot be empty.") => array has a non zero length
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/IEEE754rUtils.java]: Instead of suppressing this warning, I would annotate the argument to this method to express this fact. This follows the principle of ["annotate normal behavior"](https://checkerframework.org/manual/#annotate-normal-behavior). Everywhere that you've suppressed a warning with a message about a call to `Validate.isTrue`, you can probably add an annotation instead. For example, in this method you'll need an `@MinLen(1)` annotation on the parameter.

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java [hunk 24]
```diff
@@ -682,7 +686,7 @@ public static Number createNumber(final String str) {
             }
             return createInteger(str);
         }
-        final char lastChar = str.charAt(str.length() - 1);
+        final char lastChar = str.charAt(str.length() - 1); // #1
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: Why are these three code locations marked?

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java [hunk 25]
```diff
@@ -1013,6 +1021,7 @@ public static long min(final long... array) {
         validateArray(array);
 
         // Finds and returns min
+        @SuppressWarnings("index:array.access.unsafe.high.constant") // validateArray(array) => array.length != 0
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: Why not annotate `validateArray` as `@EnsuresMinLenIf`?

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java [hunk 26]
```diff
@@ -1599,7 +1619,8 @@ public static boolean isCreatable(final String str) {
         if (StringUtils.isEmpty(str)) {
             return false;
         }
-        final char[] chars = str.toCharArray();
+        @SuppressWarnings("value:assignment.type.incompatible") // !StringUtils.isEmpty(str) => @MinLen(1)
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: Similarly, you could annotate `StringUtils.isEmpty` as `@EnsuresMinLenIf`

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java (hunk 11 above)
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:167]: My comment here was in error. I thought that `isVarArgs` was a method that was called on `srcArgs` for some reason, which it clearly isn't. You can disregard this comment.

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java [hunk 27]
```diff
@@ -260,12 +266,13 @@ private static boolean isMatchingExecutable(final Executable method, final Class
         }
 
         if (method.isVarArgs()) {
-            int i;
+            @NonNegative int i; // annotating as @NonNegative or else it is detected as @LowerBoundBottom by the checker
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:269]: Is there an issue you can link to here where you reported this? IIRC, it might have been https://github.com/typetools/checker-framework/issues/2505, which was fixed two weeks ago - which would mean we should be able to remove both this annotation and the comment.

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 28]
```diff
@@ -95,7 +104,7 @@
     /** Internal data storage. */
     protected char[] buffer; // TODO make private?
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: To avoid warnings later in the class, this should probably be `@SameLen("this")`. See https://checkerframework.org/manual/#index-annotating-fixed-size

File: src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java [hunk 29]
```diff
@@ -119,7 +127,8 @@ public final void translate(final CharSequence input, final Writer out) throws I
      * @return CharSequenceTranslator merging this translator with the others
      */
     public final CharSequenceTranslator with(final CharSequenceTranslator... translators) {
-        final CharSequenceTranslator[] newArray = new CharSequenceTranslator[translators.length + 1];
+        @SuppressWarnings("value:assignment.type.incompatible") // translators.length + 1 has minimum value 1
+        final CharSequenceTranslator @MinLen(1) [] newArray = new CharSequenceTranslator[translators.length + 1];
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java:131]: This looks like a bug in the Value Checker. Can you write a short test case and submit a bug?

File: src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java (hunk 20 above)
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java]: Can you address the discrepancy here?

File: src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java [hunk 30]
```diff
@@ -46,15 +52,19 @@
      *
      * @param lookup CharSequence[][] table of size [*][2]
      */
-    public LookupTranslator(final CharSequence[]... lookup) {
+    @SuppressWarnings({"value:enhancedfor.type.incompatible", "index:argument.type.incompatible"}) /*
+    #1: CharSequence is of the type [*][2] (according to the documentation), hence Charsequence has length 2
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java]: Why not just write that annotation, then? You can write a `MinLen`/`ArrayLen` annotation on an inner array.

**Review by kelloggm — CHANGES_REQUESTED** [review]
There is still one issue from a previous review that hasn't been addressed. I also noticed some other warning suppressions that are only present because some methods should be annotated to reflect their normal (not exceptional) behavior.

File: pom.xml (hunk 22 above)
Comment (Kakarot-SSJ4) [reply pom.xml]: The checker does not run when I change the version to 2.9.0

File: src/main/java/org/apache/commons/lang3/math/NumberUtils.java (hunk 25 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: `validateArray` does not return a boolean type, hence I cannot annotate it with `@EnsuresMinLenIf`

Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/math/NumberUtils.java]: In that case, you should probably just annotate the argument to this method as `@MinLen(1)`, since you should [annotate normal behavior](https://checkerframework.org/manual/#annotate-normal-behavior).

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java (hunk 27 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:269]: The checker shows lowerbound:array.access.unsafe.low when I remove the annotation.

File: src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java (hunk 30 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java]: I have already annotated `seq` with `@ArrayLen(2)`

Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java]: Yes, but a warning is issued there because the typechecker can't enforce that annotation. I was suggesting writing an annotation on `lookup` that indicates that its elements are `@ArrayLen(2)`. You've currently annotated it as if its elements are `@MinLen(1)`, instead, for some reason.

File: src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java (hunk 27 above)
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/reflect/MemberUtils.java:269]: Can you write a test case and report a bug, then?

File: pom.xml (hunk 22 above)
Comment (kelloggm) [reply pom.xml]: Can you be more precise about what goes wrong? There's no reason that this shouldn't work on 2.9.0, and if it doesn't then that's something I would like to debug.

Comment (Kakarot-SSJ4) [reply pom.xml]: I might have made some mistake when I tried, it does work.
